File: server/routes/security.py
"""보안 상태 읽기 전용 조회 — 설정 → 「보안」 탭(60-settings-palette.md §2).

**읽기만 한다.** 비밀번호 설정·OTP 활성화·기기 폐기는 전부 `fsh` CLI 몫이다.
공개 터널 너머에서 인증 수단 자체를 바꿀 수 있게 만드는 건, 그 인증이 지키려던
것을 그 인증으로 들어와서 풀 수 있다는 뜻이라 설계상 배제한다(2.0 결정 유지).

두 경로 모두 AuthMiddleware 예외 목록에 없다 — 로그인한 세션만 본다.
"""

from __future__ import annotations

from fastapi import APIRouter, Request

import auth

router = APIRouter()

# 기기 id는 sha256 해시의 앞 16자다. 해시 자체로는 쿠키를 만들 수 없지만,
# 화면에 식별용으로 필요한 건 사람이 목록에서 한 행을 골라낼 만큼뿐이라
# 더 짧게 자른다(`fsh device revoke`도 접두사 매칭이라 그대로 쓸 수 있다).
_ID_DISPLAY_LEN = 8


@router.get("/api/devices")
async def list_devices(request: Request):
    """등록 기기 목록. `fsh device list`와 같은 저장소(~/.vt/devices.json)를 읽는다."""
    current = auth.verify_device(request.cookies.get("vt_device", "")) or {}
    current_id = current.get("id", "")
    devices = []
    for d in auth.list_devices():
        did = str(d.get("id", ""))
        devices.append({
            "id": did[:_ID_DISPLAY_LEN],
            "label": d.get("label", ""),
            "added_at": int(d.get("added_at", 0) or 0),
            "last_seen": int(d.get("last_seen", 0) or 0),
            "current": bool(did and did == current_id),
        })
    return {"devices": devices}


@router.get("/api/auth/elevation")
async def elevation_status(request: Request):
    """현재 세션의 승격(N31) 상태.

    ADR-27로 push/PR을 만들지 않기로 하면서 승격을 요구하는 경로가 사라졌다.
    기능은 살아 있지만 지금은 아무도 쓰지 않는다 — 화면이 "항상 비활성"으로만
    보이면 고장으로 오해되므로 그 사실(`unused`)을 응답에 같이 실어 보낸다.
    """
    until = auth.session_elevated_until(request.cookies.get("vt_session", ""))
    return {
        "elevated": bool(until),
        "elevated_until": int(until),
        "unused": True,
        "unused_reason": "ADR-27",
    }

File: frontend/js/panels/settings/sections/security.js
import { vtFetch } from '../../../core/api.js';
import { statusLine, secRow, fmtWhen } from '../controls.js';

// ── 「보안」 ──────────────────────────────────────────────────────────────
// 60-settings-palette.md §2 — 비밀번호 · OTP · 기기 · 승격 상태.
//
// **전부 읽기 전용이다.** 비밀번호 설정·OTP 활성화·기기 폐기는 `fsh` CLI에서만
// 한다. 공개 터널 너머에서 인증 수단 자체를 바꿀 수 있으면, 그 인증으로 들어온
// 사람이 그 인증을 풀 수 있다는 뜻이라 여기서는 상태와 명령어만 안내한다.
export function renderSecuritySection() {
  const frag = document.createDocumentFragment();

  const authHost = document.createElement('div');
  authHost.className = 'vt-set-sechost';
  authHost.appendChild(statusLine('인증 상태 확인 중…'));
  frag.appendChild(authHost);

  const devTitle = document.createElement('div');
  devTitle.className = 'vt-set-label';
  devTitle.textContent = '등록 기기';
  frag.appendChild(devTitle);

  const devHost = document.createElement('div');
  devHost.className = 'vt-set-sechost';
  devHost.appendChild(statusLine('기기 목록 확인 중…'));
  frag.appendChild(devHost);

  const elevHost = document.createElement('div');
  elevHost.className = 'vt-set-sechost';
  elevHost.appendChild(statusLine('승격 상태 확인 중…'));
  frag.appendChild(elevHost);

  vtFetch('/api/auth/status').then((r) => {
    authHost.innerHTML = '';
    const pw = !!(r && r.password_set);
    authHost.appendChild(secRow(
      '웹 로그인 비밀번호',
      pw ? '설정됨' : '설정 안 됨',
      pw ? 'on' : 'off',
      pw ? null : "터미널에서 'fsh password'로 설정합니다.",
    ));
    const otp = !!(r && r.otp_enabled);
    authHost.appendChild(secRow(
      'OTP (새 기기 등록 관문)',
      otp ? '활성' : '비활성',
      otp ? 'on' : 'off',
      otp ? '처음 보는 기기를 등록할 때만 OTP를 요구합니다.'
          : "터미널에서 'fsh otp setup'으로 활성화합니다.",
    ));
  }).catch(() => { authHost.innerHTML = ''; authHost.appendChild(statusLine('인증 상태를 확인할 수 없습니다.')); });

  vtFetch('/api/devices').then((r) => {
    const rows = (r && Array.isArray(r.devices)) ? r.devices : [];
    devHost.innerHTML = '';
    if (!rows.length) {
      devHost.appendChild(statusLine('등록된 기기가 없습니다. 휴대폰에서 QR로 처음 접속하면 등록됩니다.'));
      return;
    }
    devHost.appendChild(renderDeviceTable(rows));
    devHost.appendChild(statusLine("기기 폐기는 터미널에서 'fsh device revoke <id>'로 합니다 — 그 기기의 세션도 함께 끊깁니다."));
  }).catch(() => { devHost.innerHTML = ''; devHost.appendChild(statusLine('기기 목록을 확인할 수 없습니다.')); });

  vtFetch('/api/auth/elevation').then((r) => {
    const until = (r && Number(r.elevated_until)) || 0;
    const left = until ? Math.max(0, Math.round((until * 1000 - Date.now()) / 60000)) : 0;
    elevHost.innerHTML = '';
    elevHost.appendChild(secRow(
      '승격 세션',
      until && left ? `승격됨 · ${left}분 남음` : '승격 안 됨',
      until && left ? 'on' : 'off',
      // ADR-27 — push/PR을 만들지 않기로 해서 승격을 요구하는 경로가 지금은 없다.
      // 이 문장이 없으면 "왜 항상 비활성이지?"로 읽힌다.
      '현재 이 기능을 쓰는 경로가 없습니다(ADR-27) — 승격을 요구하는 조작이 아직 없습니다.',
    ));
  }).catch(() => { elevHost.innerHTML = ''; elevHost.appendChild(statusLine('승격 상태를 확인할 수 없습니다.')); });

  return frag;
}

function renderDeviceTable(rows) {
  const table = document.createElement('table');
  table.className = 'vt-set-covtable vt-set-devtable';

  const thead = document.createElement('thead');
  const headRow = document.createElement('tr');
  for (const label of ['기기', 'id', '등록', '마지막 사용']) {
    const th = document.createElement('th');
    th.textContent = label;
    headRow.appendChild(th);
  }
  thead.appendChild(headRow);
  table.appendChild(thead);

  const tbody = document.createElement('tbody');
  for (const d of rows) {
    const tr = document.createElement('tr');
    if (d.current) tr.dataset.current = '1';

    const label = document.createElement('td');
    label.textContent = d.label || '기기';
    if (d.current) {
      const badge = document.createElement('span');
      badge.className = 'vt-tag vt-set-devme';
      badge.textContent = '이 기기';
      label.appendChild(badge);
    }

    const id = document.createElement('td');
    id.className = 'vt-set-covcli';
    id.textContent = d.id || '—';

    const added = document.createElement('td');
    added.textContent = fmtWhen(d.added_at);

    const seen = document.createElement('td');
    seen.textContent = fmtWhen(d.last_seen);

    tr.append(label, id, added, seen);
    tbody.appendChild(tr);
  }
  table.appendChild(tbody);
  return table;
}

